src/cast_core/src/AssertiveActor.cpp

```diff
@@ -0,0 +1,216 @@
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/DocumentGenerator.hpp>
+
+namespace genny::actor {
+
+
+struct AssertiveActor::PhaseConfig {
+    mongocxx::database database;
+    DocumentGenerator expectedExpr;
+    DocumentGenerator actualExpr;
+    std::optional<std::string> message;
+    std::set<std::string> ignoreFields;
+
+    PhaseConfig(PhaseContext& phaseContext, mongocxx::database&& db, ActorId id)
+        : database{db},
+          expectedExpr{phaseContext["Expected"].to<DocumentGenerator>(phaseContext, id)},
+          actualExpr{phaseContext["Actual"].to<DocumentGenerator>(phaseContext, id)},
+          message{phaseContext["Message"].maybe<std::string>()} {
+        for (auto [k, fieldNode] : phaseContext["IgnoreFields"]) {
+            ignoreFields.insert(fieldNode.to<std::string>());
+        }
+    }
+};
+
+const std::string prefix(AssertiveActor::PhaseConfig* config) {
+    return "Assert" +(config->message ? " \"" + *(config->message) + "\" ": "") + ": ";
+}
+
+auto runCommandAndGetResult(AssertiveActor::PhaseConfig* config, bsoncxx::document::value& command) {
+    auto cmdView = command.view();
+
+    BOOST_LOG_TRIVIAL(info) << prefix(config) << bsoncxx::to_json(cmdView);
+    auto res = config->database.run_command(std::move(command));
+    auto resView = res.view();
+
+    return res;
+}
+
+auto getBatchFromCommandResult(const bsoncxx::document::view& resView) {
+    if (resView["ok"]) {
+        auto cursor = resView["cursor"].get_document().view();
+        auto firstBatch = cursor["firstBatch"].get_array().value;
+        return firstBatch;
+    }
+    BOOST_THROW_EXCEPTION(AssertFailed("Failed to run command."));
+}
+
+template <typename T>
+double convertNumericToDouble(const T& numeric) {
+    switch(numeric.type()) {
+        case bsoncxx::type::k_double:
+            return numeric.get_double();
+        case bsoncxx::type::k_int32:
+            return numeric.get_int32() * 1.0;
+        case bsoncxx::type::k_int64:
+            return numeric.get_int64() * 1.0;
+        default:
+            BOOST_THROW_EXCEPTION(AssertFailed("not a numeric"));
+    }
+}
+
+bool isNumeric(bsoncxx::type type) {
+    return (type == bsoncxx::type::k_double)
+        || (type == bsoncxx::type::k_int32)
+        || (type == bsoncxx::type::k_int64);
+}
+
+bool equalBSONDocs(AssertiveActor::PhaseConfig* config, const bsoncxx::document::view& expected, const bsoncxx::document::view& actual);
+bool equalBSONArrays(AssertiveActor::PhaseConfig* config, const bsoncxx::array::view& expected, const bsoncxx::array::view& actual);
+
+template <typename T>
+bool equal(AssertiveActor::PhaseConfig* config, const T& expectedVal, const T& actualVal) {
+    auto type = expectedVal.type();
+    if (isNumeric(type) && isNumeric(actualVal.type())) {
+        // Rounding may result in one value being a double while another is integral. Handle this case through double comparison.
```

Comment:
Where does this rounding occur? It seems like the same docs would be returned from either query and still be directly comparable for each type?

(I notice the C++ driver tests do the same thing so it's undoubtedly correct, I'm just curious about what I'm missing!)